File: derivatives/validator-rs/src/chainstate.rs
//! The stateful validator's UTXO + value rules — mirrors the Python `ChainState._connect`. NOT money.
//!
//! `apply_txs` is the shared value/script logic (used by the simple `ChainState` here and by the
//! reorg‑capable `NodeState` in `reorg`): every input exists and is unspent, coinbase maturity, the
//! input script is satisfied (the full v0.1 interpreter, `script::verify_spend`), no inflation, and
//! the coinbase‑value rule with fees — applied with rollback so a rejected block leaves the UTXO
//! unchanged.

use std::collections::HashMap;

use crate::script::verify_spend;
use crate::{check_coinbase_value, is_coinbase, parse_block_txs};

pub type Outpoint = ([u8; 32], u32);

/// An unspent output.
#[derive(Clone)]
pub struct Coin {
    pub value: i64,
    pub spk: Vec<u8>,
    pub height: i64,
    pub coinbase: bool,
}

/// Undo data for one connected block: (pre‑block coins consumed, outputs created).
pub type Undo = (Vec<(Outpoint, Coin)>, Vec<Outpoint>);

/// Apply a block's transactions to `utxo`, enforcing the value rules. On success returns the undo
/// data (UTXO mutated); on the first failing rule it **rolls back** and returns the reason.
pub(crate) fn apply_txs(
    utxo: &mut HashMap<Outpoint, Coin>,
    raw: &[u8],
    height: i64,
    subsidy: i64,
    strict: bool,
    maturity: i64,
    is_genesis: bool,
) -> Result<Undo, &'static str> {
    let txs = parse_block_txs(raw);
    let mut created: Vec<Outpoint> = Vec::new();
    let mut spent_prior: Vec<(Outpoint, Coin)> = Vec::new();
    let mut fees: i64 = 0;

    macro_rules! bail {
        ($e:expr) => {{
            for k in &created {
                utxo.remove(k);
            }
            for (k, c) in spent_prior.into_iter().rev() {
                utxo.insert(k, c);
            }
            return Err($e);
        }};
    }

    for (tx, tid) in &txs {
        let coinbase = is_coinbase(tx);
        if !coinbase {
            let mut value_in: i64 = 0;
            for (i_in, vin) in tx.vin.iter().enumerate() {
                let key = (vin.prevhash, vin.n);
                let coin = match utxo.get(&key) {
                    Some(c) => c.clone(),
                    None => bail!("input missing or already spent"),
                };
                if coin.coinbase && height - coin.height < maturity {
                    bail!("immature coinbase spend");
                }
                if !verify_spend(&vin.script, &coin.spk, tx, i_in) {
                    bail!("input script does not satisfy output");
                }
                value_in += coin.value;
                utxo.remove(&key);
                if let Some(pos) = created.iter().position(|k| *k == key) {
                    created.remove(pos); // same-block output consumed -> nets out
                } else {
                    spent_prior.push((key, coin));
                }
            }
            let value_out: i64 = tx.vout.iter().map(|o| o.value).sum();
            if value_in < value_out {
                bail!("inflation (inputs < outputs)");
            }
            fees += value_in - value_out;
        }
        for (n, o) in tx.vout.iter().enumerate() {
            let k = (*tid, n as u32);
            utxo.insert(k, Coin { value: o.value, spk: o.script.clone(), height, coinbase });
            created.push(k);
        }
    }

    if !is_genesis {
        let claimed: i64 = txs[0].0.vout.iter().map(|o| o.value).sum();
        if !check_coinbase_value(claimed, subsidy, fees, strict) {
            bail!("coinbase value violates the chain rule");
        }
    }
    Ok((spent_prior, created))
}

/// A single-chain UTXO validator (no reorg) — connect a block directly. For reorg + difficulty use
/// [`crate::reorg::NodeState`].
pub struct ChainState {
    utxo: HashMap<Outpoint, Coin>,
    maturity: i64,
    strict: bool,
    pub height: i64,
}

impl ChainState {
    pub fn new(maturity: i64, strict: bool) -> Self {
        ChainState { utxo: HashMap::new(), maturity, strict, height: -1 }
    }

    pub fn utxo_count(&self) -> usize {
        self.utxo.len()
    }

    pub fn balance(&self) -> i64 {
        self.utxo.values().map(|c| c.value).sum()
    }

    /// Connect `raw` at `height` (atomically). `Ok(())` on success, or the first failing reason
    /// (UTXO unchanged).
    pub fn connect_block(
        &mut self,
        raw: &[u8],
        height: i64,
        subsidy: i64,
        is_genesis: bool,
    ) -> Result<(), &'static str> {
        apply_txs(&mut self.utxo, raw, height, subsidy, self.strict, self.maturity, is_genesis)?;
        self.height = height;
        Ok(())
    }
}

File: derivatives/netnode/chainstate.py
"""A validated UTXO chainstate — Path B toward a production node (part 2).

Layered over the chainsync block index (PoW tree), this maintains the **active, fully‑validated
chain** and its **UTXO set**, with reorg‑safe **connect / disconnect** (undo data per block) and
`activate_best`, which moves the validated chain toward the index's best chain — **gating every
step on full validity** and rolling back (restoring the prior chain) if a branch fails to validate.

Per block it enforces, on top of the structural/PoW checks (`fullnode.validate_block`):
- **difficulty** — `nBits` equals the expected retarget for the parent (floored at `min_bits`).
  Because this authoritative gate runs on *connect*, it also covers the **orphan reconnection
  path**, where the direct‑receipt difficulty check is deferred (parent not yet known),
- every input exists in the UTXO (no missing / double‑spent coins),
- coinbase maturity (`COINBASE_MATURITY`),
- the input script satisfies the referenced output's script (the v0.1 `VerifySignature` path,
  full vocabulary — via `model/spend.verify_spend`),
- no inflation (Σ inputs ≥ Σ outputs), and
- the chain's coinbase‑value rule with fees (`Rules.coinbase_ok`).

Same‑block spends (a later tx spending an earlier tx's output) are handled, and the undo nets
them out. Evidence: MODEL / NEW‑EXP.
"""

from __future__ import annotations

import pathlib
import sys

_HERE = pathlib.Path(__file__).resolve().parent
for _p in ("model", "p2p", "nov08x"):
    sys.path.insert(0, str(_HERE.parent / _p))
sys.path.insert(0, str(_HERE))

import cscript                                              # noqa: E402

from difficulty import expected_bits                       # noqa: E402
from fastverify import verify_spend_fast                    # noqa: E402  (== faithful verify_spend, accelerated)
from fullnode import is_coinbase, parse_block_with_txids   # noqa: E402

COINBASE_MATURITY = 100


class InvalidBlock(Exception):
    pass


class Coin:
    __slots__ = ("value", "spk", "height", "coinbase")

    def __init__(self, value, spk, height, coinbase):
        self.value, self.spk, self.height, self.coinbase = value, spk, height, coinbase


class ChainState:
    def __init__(self, chain, rules, maturity: int = COINBASE_MATURITY, min_bits: int | None = None,
                 reopen=frozenset()):
        self.chain = chain
        self.rules = rules
        self.maturity = maturity
        self.min_bits = min_bits                            # difficulty floor (None -> genesis nBits)
        self.reopen = frozenset(reopen)                     # script posture: {} faithful | {'OP_NOTEQUAL'} nothing-disabled
        self.utxo: dict[tuple[bytes, int], Coin] = {}
        self.active: list[bytes] = []                       # genesis .. validated tip
        self.undo: dict[bytes, tuple[list, list]] = {}      # block -> (spent_prior, created_keys)
        self.invalid: set[bytes] = set()
        if chain.genesis is not None:
            self._connect(chain.genesis)                    # genesis coinbase seeds the UTXO

    # -- queries ---------------------------------------------------------------
    @property
    def tip(self) -> bytes | None:
        return self.active[-1] if self.active else None

    @property
    def height(self) -> int:
        return self.chain.by_hash[self.active[-1]].height if self.active else -1

    def balance(self) -> int:
        return sum(c.value for c in self.utxo.values())

    # -- locator / serving over the VALIDATED chain (authoritative) -------------
    def get_locator(self) -> list[bytes]:
        have, i, step = [], len(self.active) - 1, 1
        while i >= 0:
            have.append(self.active[i])
            i -= step
            if len(have) > 10:
                step *= 2
        if self.active and have[-1] != self.active[0]:
            have.append(self.active[0])
        return have

    def blocks_after(self, have, hash_stop) -> list[bytes]:
        pos = {h: k for k, h in enumerate(self.active)}
        start = max((pos[h] for h in have if h in pos), default=0)
        out = []
        for h in self.active[start + 1:]:
            if h == hash_stop:
                break
            out.append(h)
        return out

    # -- connect / disconnect --------------------------------------------------
    def _connect(self, h: bytes):
        idx = self.chain.by_hash[h]
        height = idx.height
        txs = parse_block_with_txids(idx.raw)                # [(tx, txid)] — txid from parsed bytes
        is_genesis = h == self.chain.genesis
        if not is_genesis and idx.nBits != expected_bits(self.chain, idx.prev, self.rules, self.min_bits):
            raise InvalidBlock("wrong difficulty")           # authoritative — also covers orphan reconnection
        created: dict[tuple[bytes, int], Coin] = {}          # block-created outputs still live
        spent_prior: list = []                               # (outpoint, coin) pre-block coins consumed
        fees = 0
        try:
            for tx, tid in txs:
                coinbase = is_coinbase(tx)
                if not coinbase:
                    value_in = 0
                    for i, vin in enumerate(tx.vin):
                        key = (vin.prevhash, vin.n)
                        coin = self.utxo.get(key)
                        if coin is None:
                            raise InvalidBlock("input missing or already spent")
                        if coin.coinbase and height - coin.height < self.maturity:
                            raise InvalidBlock("immature coinbase spend")
                        if not verify_spend_fast(cscript.parse(vin.script), coin.spk, tx, i, reopen=self.reopen):
                            raise InvalidBlock("input script does not satisfy output")
                        value_in += coin.value
                        del self.utxo[key]
                        if key in created:
                            del created[key]                 # same-block output consumed -> nets out
                        else:
                            spent_prior.append((key, coin))
                    value_out = sum(o.value for o in tx.vout)
                    if value_in < value_out:
                        raise InvalidBlock("inflation (inputs < outputs)")
                    fees += value_in - value_out
                for n, o in enumerate(tx.vout):
                    k = (tid, n)
                    c = Coin(o.value, cscript.parse(o.script), height, coinbase)
                    self.utxo[k] = c
                    created[k] = c
            if not is_genesis:
                subsidy = self.rules.get_block_value(height - 1)
                claimed = sum(o.value for o in txs[0][0].vout)
                if not self.rules.coinbase_ok(claimed, subsidy + fees):
                    raise InvalidBlock("coinbase value violates the chain rule")
        except InvalidBlock:                                 # atomic: revert partial mutations
            for k in created:
                self.utxo.pop(k, None)
            for key, coin in reversed(spent_prior):
                self.utxo[key] = coin
            raise
        self.undo[h] = (spent_prior, list(created.keys()))
        self.active.append(h)

    def _disconnect(self):
        h = self.active.pop()
        spent_prior, created_keys = self.undo.pop(h)
        for k in created_keys:
            self.utxo.pop(k, None)
        for key, coin in spent_prior:
            self.utxo[key] = coin

    # -- activate the best VALID chain (reorg-safe, gated on validity) ---------
    def activate_best(self):
        target = self.chain.main_chain()                    # the index's best-by-work chain
        if self.active and self.active[-1] == target[-1]:
            return
        old_active = list(self.active)
        old_height = self.height
        fork = 0
        while (fork < len(old_active) and fork < len(target)
               and old_active[fork] == target[fork]):
            fork += 1
        while len(self.active) > fork:                       # disconnect down to the fork
            self._disconnect()
        for h in target[fork:]:                              # connect the new branch, validating
            if h in self.invalid:
                break
            try:
                self._connect(h)
            except InvalidBlock:
                self.invalid.add(h)
                break
        if self.height <= old_height and self.active != old_active:
            while len(self.active) > fork:                   # reorg didn't improve -> restore old chain
                self._disconnect()
            for h in old_active[fork:]:
                self._connect(h)                             # previously valid -> reconnects
